utils/deploy_to_dev_env.sh

```diff
@@ -4,55 +4,120 @@
-if [ -z "$ENVIRONMENT" ]; then
-  echo "You must enter a non-empty string"
-  exit 1
-fi
+function print_error() {
+  printf "\t$0: Error: %s\n\n" "$*" >&2;
+}
 
-STACK_NAME="core-front-${ENVIRONMENT}"
-DEV_IMAGE_TAG="${ENVIRONMENT}-$(date +%s)"
+function process_args() {
+  if [[ $# -lt 1 ]]; then
+    print_usage
+  fi
 
-if ! aws cloudformation describe-stacks \
-    --stack-name "$STACK_NAME" \
-    --region eu-west-2 > /dev/null 2>&1 ; then
-  echo "The stack 'core-front-${ENVIRONMENT}' has not been found."
-  echo " - check the script has been run with the required permissions, e.g. aws-vault exec <profile> -- $0"
-  echo " - check you are on the VPN"
-  echo " - check whether your stack exists via the console or cli, if not you may need to: "
-  echo "    - Run your developer pipeline in Concourse to create it"
-  echo "    - if you do not have a developer pipeline then ask a team member to update di-ipv-config/core/ci/core-developer-pipelines"
-  exit 1
-fi
+  while getopts fhx:i:e: flag; do
+    case $flag in
+      e) ENVIRONMENT=${OPTARG}
+      ;;
+      h) print_usage && exit 0
+      ;;
+      \?)
+        print_error "Unrecognized argument '${OPTARG}'"
+      ;;
+      *) print_usage >&2
+         exit 1
+      ;;
+    esac
+  done
+
+  [ -z "$ENVIRONMENT" ] || export ENVIRONMENT=$ENVIRONMENT
+  [ -z "$ENVIRONMENT" ] && print_error "Environment Name: ${ENVIRONMENT} not provided (-e) :: exiting" && exit 1
+}
 
-IMAGE_NAME_AND_TAG="core-front:${DEV_IMAGE_TAG}"
+function check_dependencies() {
+  if ! which jq >/dev/null; then
+    print_error "Please install jq to use this script - more info here: https://formulae.brew.sh/formula/jq"
+    exit 1
+  fi
 
+  if ! docker info > /dev/null 2>&1 ; then
+    print_error "Docker engine is not running, please start it. more info here: https://www.docker.com/products/docker-desktop/"
+    exit 1
+  fi
+}
 
-docker build -t "$IMAGE_NAME_AND_TAG" ../
+function check_connection() {
+
+  if ! aws cloudformation describe-stacks \
+      --stack-name "$STACK_NAME" \
+      --region "${region}" > /dev/null 2>&1 ; then
+    echo "The stack 'core-front-${ENVIRONMENT}' has not been found in region: '${region}'."
+    echo " - check the script has been run with the required permissions, e.g. aws-vault exec <profile> -- $0"
+    echo " - check you are on the VPN"
+    echo " - check whether your stack exists via the console or cli, if not you may need to: "
+    echo "    - Run your developer pipeline in Concourse to create it"
+    echo "    - if you do not have a developer pipeline then ask a team member to update di-ipv-config/core/ci/core-developer-pipelines"
+    exit 1
+  fi
+}
 
-echo "Logging into developement ECR registry"
-ecr_registry="130355686670.dkr.ecr.eu-west-2.amazonaws.com"
-aws ecr get-login-password --region eu-west-2 | \
-  docker login --username AWS --password-stdin "${ecr_registry}"
+function clone_config_repo() {
+  git clone git@github.com:alphagov/di-ipv-config.git config-repo
```

Comment:
So at present if I do not provide the `-e` flag the error message isn't right because process_args is called before init and the `environment` array is not populated, not a huge deal but I imagine it'll happen frequently the first time people run it. This is what you get
```
GDS11321:utils dan.worth$ aws-vault exec di-ipv-dev -- ./deploy_to_dev_env.sh

Usage: ./deploy_to_dev_env.sh:
        [-e] Provide environment name from:
```


I get why you're cloning the repo but ideally I don't want to wait to clone it when I already have it (its also unexpected). For simplicity why not just add a link to the file in the help text, something like the following and add anything extra to the README? Or look for it on the dev's laptop, most people will clone it into the same parent directory. I feel like we're on the brink of over engineering this, despite my fondness for a bit of `bash` :)

```
echo -e "\t[-e] Provide environment name from those listed at https://github.com/alphagov/di-ipv-config/blob/main/core/ci/core-developer-pipelines/generate-pipelines/list_of_developers.txt\n"
```

Reply:
If you want to be helpful and provide a list of configured envs then how about using something like the following to derive it from the available front end stacks?? This also means it'll make sure the stack actually exists which is necessary further on....
```
GDS11321:utils dan.worth$ aws-vault exec dev-admin -- aws cloudformation list-stacks --region eu-west-2 | jq '.StackSummaries[] | select(.StackName | contains("front")).StackName' -r | sort | uniq | cut -d '-' -f 3,4
dev-amrits
dev-aprilm
dev-chrisw
dev-danh
dev-danp
dev-danw
dev-lait
dev-patrickb
dev-vamh
dev-wojciecho
```

Reply:
To be honest I'm leaning towards a link to the file and a note that if they're not listed to ask for their environment to be set up. That seems the most simple but open to other ideas.